src/data_source.rs

```diff
@@ -260,70 +265,163 @@ where
     }
 }
 
-#[derive(Clone, Copy, Debug, Snafu)]
-#[snafu(display("unable to open stream"))]
-pub struct StreamError;
+async fn range_stream<T>(
+    mut iter: Iter<'_, T>,
+    range: impl RangeBounds<usize>,
```

Comment:
Why make this an `impl RangeBounds<usize>` instead of a `Range<usize>`?

Reply:
Because `from..to` is a `Range`, but `..to` is a `RangeTo`, and there are also `RangeFrom` and some others. So the way to deal with all types of ranges is to take `RangeBounds`. It's kind of annoying, I sort of think they should all be the same type of struct with optional lower and upper bounds, but that's how std does it